custom_metrics/pwa.js

```diff
@@ -69,17 +69,33 @@ const manifests = getEntriesForURLs(manifestURLs).map(([url, body]) => {
 const serviceWorkerInitiatedURLs = new Set(Array.from(serviceWorkerURLs).flatMap(getURLsInitiatedBy));
 const serviceWorkerInitiated = getEntriesForURLs(serviceWorkerInitiatedURLs);
 
-const workboxPattern = /(?:workbox:[a-z\-]+:[\d.]+|workbox\.[a-zA-Z]+\.?[a-zA-Z]*)/g;
 // We should use serviceWorkerInitiatedURLs here, but SW detection has some false negatives.
-const workboxInfo = response_bodies.filter(har => {
-  return workboxPattern.test(har.response_body);
-}).map(har => {
-  return [har.url, Array.from(har.response_body.matchAll(workboxPattern)).map(m => m[0])];
-});
+function getInfoForPattern(regexPattern) {
+  return response_bodies.filter(har => {
+    return regexPattern.test(har.response_body);
+  }).map(har => {
+    return [har.url, Array.from(har.response_body.matchAll(regexPattern)).map(m => m[0])];
+  });
+}
+
+const workboxPattern = /(?:workbox:[a-z\-]+:[\d.]+|workbox\.[a-zA-Z]+\.?[a-zA-Z]*)/g;
+const workboxInfo = getInfoForPattern(workboxPattern);
+
+const importScriptsPattern = /importScripts\(.*\);/g;
+const importScriptsInfo = getInfoForPattern(importScriptsPattern);
+
+const swEventListenersPattern = /addEventListener\(\s*[\'"](install|activate|fetch|push|notificationclick|notificationclose|sync|canmakepayment|paymentrequest|message|messageerror|periodicsync|backgroundfetchsuccess|backgroundfetchfailure|backgroundfetchabort|backgroundfetchclick)[\'"]/g;
+const swEventListenersInfo = getInfoForPattern(swEventListenersPattern);
+
+const swPropertiesPattern = /\.on(install|activate|fetch|push|notificationclick|notificationclose|sync|canmakepayment|paymentrequest|message|messageerror|periodicsync|backgroundfetchsuccess|backgroundfetchfailure|backgroundfetchabort|backgroundfetchclick)\s*=/g;
+const swPropertiesInfo = getInfoForPattern(swPropertiesPattern);
```

Comment:
Do you need to distinguish between `addEventListener` and `on` usage in your analysis? If not, consider combining the regexes or merging the results into a single output property.

Also FYI we have the new [event-names.js](https://github.com/HTTPArchive/legacy.httparchive.org/blob/master/custom_metrics/event-names.js) custom metric that will extract any event name after `addEventListener`. Probably still good to have a shortlist of SW-related event names to check for here but just wanted to make you aware in case it's useful down the road.

Reply:
We decided not to combine fields, since they are easier to merge during queries, but harder to separate, if, for some reason we need to use the original separation. In this case, for example, to answer the question: "How many sites use event listeners vs. properties".
Please, refer to the **Granularity level** headline in [this comment](https://github.com/HTTPArchive/legacy.httparchive.org/pull/204#issuecomment-858998233) for more details.